### PR TITLE
Add Module Operation commands for newest PLC chipsets for HomePlug AV + HomePlug GREENPHY and Smart Cities (SG) layers

### PR DESCRIPTION
**Pull request by FlUxIuS**
This PR adds Module Operation that could be found in new PLC QCA chipsets and a HomePlug GREENPHY layer to play with cars, charging stations and other devices using this technology.

### CODE REVIEW

**Comment by guedou** [comment]
Thanks for your PR.

Your code has several PEP08 related errors, see https://travis-ci.com/secdev/scapy/jobs/241671466#L660. Locally you can easily reproduce them by installing the `tox` tool and invoking `tox -e flake8`.

**Comment by FlUxIuS** [comment]
Thanks Guillaume!

Seems like I have another problem, but doesn't looks related to my modification (directly) ^^

**Comment by gpotter2** [comment]
Hi !
I restarted the failing tests. Could you please add some unit tests ?
https://scapy.readthedocs.io/en/latest/development.html#testing-with-utscapy
Thanks !

File: test/contrib/homeplugsg.uts [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+~ basic HomePlugAV VS_UART_CMD_REQ VS_UART_CMD_CNF
+
+HomePlugAV()/VS_UART_CMD_REQ()
+HomePlugAV()/VS_UART_CMD_CNF()
+
+= Some important manipulations
+~ field
+pkt=HomePlugAV()/VS_UART_CMD_REQ()
+pkt.UData = "AT+LOG?"
+assert str(pkt) == str(b'\x00\x00\xa4\x00\xb0R\x00\x01AT+LOG?')
```
Comment (gpotter2) [inline test/contrib/homeplugsg.uts]: To be python 3 compatible:
```
assert bytes(pkt) == b"blablabla"
```

File: scapy/contrib/homeplugav.py [hunk 2]
```diff
@@ -476,6 +476,101 @@ class ReadMACMemoryConfirmation(Packet):
...
+                                    lambda pkt:(0x0011 == pkt.operation)),
+                   ConditionalField(StrLenField("ModuleData", b"\x00",
+                                                length_from=lambda pkt: pkt.WriteDataLen),  # noqa: E501
+                                    lambda pkt:(0x0011 == pkt.operation))]
+
+    def post_build(self, p, pay):
+        if self.operation == 0x0000:
+            if self.OPDataLength is None:
+                _len = 18
+                p = p[:7] + struct.pack('h', _len) + p[9:]
```
Comment (gpotter2) [inline scapy/contrib/homeplugav.py]:
```suggestion
                p = p[:7] + struct.pack('!H', _len) + p[9:]
```
Please stick to network values (=Big endian). https://docs.python.org/2/library/struct.html#byte-order-size-and-alignment

also applies elsewhere

File: test/contrib/homeplugsg.uts [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+############
++ Basic tests
+
+* Those test are here mainly to check nothing has been broken
+
+= Building packets packet
+~ basic HomePlugAV VS_UART_CMD_REQ VS_UART_CMD_CNF
+
+HomePlugAV()/VS_UART_CMD_REQ()
+HomePlugAV()/VS_UART_CMD_CNF()
```
Comment (gpotter2) [inline test/contrib/homeplugsg.uts]: This isn't a useful test.
Good tests usually are:
- a few build tests:
```python
assert raw(IP(dst="127.0.0.1", src="127.0.0.1")) == b"\...."
assert raw(Ether()) == b"\...."
...
```
- a few dissection tests:
```python
pkt = Ether(b"....")
assert pkt.dst = ...
...
pkt = IP(...)
assert pkt.dst = ...
....
```
```

Ideally you would have one per layer you created. This also applies elsewhere

File: scapy/contrib/homeplugsg.py [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+    name = "VS_UART_CMD_CNF"
+    fields_desc = [StrFixedLenField("reserved", b"\x00", 6),
+                   FieldLenField("UDataLen", None, count_of="UData", fmt="H"),
+                   StrLenField("UData", "UartCommand\x00",
+                               length_from=lambda pkt: pkt.UDataLen)]
+
+
+# END #
+
+bind_layers(HomePlugAV, VS_UART_CMD_REQ, {"HPtype": 0xA400})
```
Comment (gpotter2) [inline scapy/contrib/homeplugsg.py]: Preferred format is
`bind_layers(A, B, HPtype=0xA400)`

Same applies elsewhere. I am aware that there is old code in Homeplugav.py that uses this. It shouldn't

File: test/contrib/homeplugav.uts [hunk 5]
```diff
@@ -48,29 +54,25 @@ HomePlugAV()/SnifferIndicate()
 pkt = HomePlugAV()/SetEncryptionKeyRequest()
 pkt.NMK = "A" * 16
 pkt.DAK = "B" * 16
-r = raw(pkt)
-r
-assert r == b"\x00P\xa0\x00\xb0R\x00AAAAAAAAAAAAAAAA\x00\xff\xff\xff\xff\xff\xffBBBBBBBBBBBBBBBB"
```
Comment (gpotter2) [inline test/contrib/homeplugav.uts]: This looks pretty unnecessary

File: scapy/contrib/homeplugsg.py (hunk 4 above)
Comment (FlUxIuS) [reply scapy/contrib/homeplugsg.py]: Done

File: test/contrib/homeplugsg.uts (hunk 3 above)
Comment (FlUxIuS) [reply test/contrib/homeplugsg.uts]: Erased + some more tests have been introduced for the HPGP protocol

File: test/contrib/homeplugsg.uts (hunk 1 above)
Comment (FlUxIuS) [reply test/contrib/homeplugsg.uts]: Done

File: scapy/contrib/homeplugav.py (hunk 2 above)
Comment (FlUxIuS) [reply scapy/contrib/homeplugav.py]: Done

File: test/contrib/homeplugav.uts (hunk 5 above)
Comment (FlUxIuS) [reply test/contrib/homeplugav.uts]: Made some modification, but not sure if this actually what you've requested.

Thanks in advance!

**Comment by evverx** [comment]
@FlUxIuS it's probably a long shot but I thought I'd ask anyway. Trying to get the scapy unit tests to pass on big-endian machines I noticed that the homeplugav module is inconsistent in terms of packs and unpacks. For example, `ModuleOperationRequest` declares the `OPDataLength` field with `LEShortField` (which implies that it's unpacked with `<H`) but `post_build` packs it with `!H` and `18` (`0x0012`) turns into `4608` (`0x1200`) below:
```sh
ModuleOperationRequest().show2()
###[ ModuleOperationRequest ]###
  reserved  = 0x0
  NumOpData = 0x1
  operation = Read
  OPDataLength= 4608
  reserved_1= 0x0
  ModuleID  = 28674
  ModuleSubID= 0
  ReadDataLen= 1400
  ReadOffset= 0
```
There are other places where unsigned short integers ('H') are packed with ('h') and so on. I'd probably fix it but unfortunately the main spec is huge and some `HPtypes` are missing there (because I assume they are vendor-specific). The wireshark homeplug dissector doesn't support all the types either. I wonder if it's fixable or should I just turn off the homeplug tests to make the other tests pass?

**Comment by FlUxIuS** [comment]
That's weird... I should check with some old and new capture to observe that and also remember what I have done! ^^'

**Pull request merged**